### PR TITLE
Upgrade Groovy to 2.5.4

### PR DESCRIPTION
**Pull request by henrik242**
Groovy 2.5.x has better jdk10 support.  

groovy-all has switched from being a jar to a pom.  Including dependencies explicitly makes it easier for Gradle to cope with that.

### CODE REVIEW

**Comment by javabrett** [comment]
This is very nice, thanks!  I'm a little perplexed by what went-on in Travis for this ... seems like it built the branch but not the `testAll` suite I expect for a PR ...

Current practice is to release from `master` (could change ...), so I guess my main question is whether the Groovy RC should be merged to `master` straight-away ... I'm not familiar with the usual cadence of their RCs and releases.  We'd probably prefer to release next with Groovy 2.5 GA, or have the option of releasing before that if it is delayed.

Unless there's a rush, maybe keep this unmerged for a week or two and we can track the Groovy release.  I'd also like to see this building properly in Travis so I might need to look into that.  #39 would also be a nice testing fix to get in, so we can start cross-compiling for newer JDKs and get some integration test coverage beyond JDK8.

**Comment by henrik242** [comment]
@javabrett Isn't it building properly in Travis?

**Comment by henrik242** [comment]
> But this PR only triggered the `/push`

@javabrett  Probably because I force pushed the branch again after some initial typos.

**Comment by javabrett** [comment]
@henrik242 I checked-out the branch and ran:

```
git commit --amend --date="$(date -R)"
git push -f origin HEAD
```
... to trigger a new Travis build, which it did, and it looks to be running both `/push` and `/pr`, good.  When they are both complete you can try the same if you like to make sure there isn't some weird Travis permissions thing.

**Comment by henrik242** [comment]
@javabrett Will do!

**Comment by javabrett** [comment]
```
> Task :gretty-integrationTests:extraResourceBases:prepareInplaceWebApp
Exception in thread "main" java.lang.NoClassDefFoundError: org/apache/commons/cli/DefaultParser
	at groovy.util.CliBuilder.parse(CliBuilder.groovy:356)
	at groovy.util.CliBuilder$parse$1.call(Unknown Source)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:47)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:116)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:128)
	at org.akhikhl.gretty.Runner.main(Runner.groovy:42)
Caused by: java.lang.ClassNotFoundException: org.apache.commons.cli.DefaultParser
	at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:335)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	... 6 more
```

**Comment by henrik242** [comment]
Hmmm! https://github.com/gretty-gradle-plugin/gretty/pull/40/commits/8d4ba37ead9d13ecdba0cde6bd943cdbe8199c9c#diff-8646794d183c19f6d0ed493cbd6a947eR10 was supposed to fix that.

**Comment by javabrett** [comment]
Looks like bits of a pre-1.3 `commons-cli` are still finding their way into the classpath, possibly from https://github.com/gretty-gradle-plugin/gretty/blob/groovy-2-5-rc-1/libs/gretty-runner/build.gradle#L6 .

**Comment by henrik242** [comment]
@javabrett Good catch, I overlooked those.  I fixed that and pushed again, let's see...

File: .travis.yml [hunk 1]
```diff
@@ -28,6 +28,8 @@ script:
   - |
     if [[ "${TRAVIS_JDK_VERSION}" == "oraclejdk8" && ( "${TRAVIS_BRANCH}" == "master" || "${TRAVIS_EVENT_TYPE}" == "pull_request" || "${TRAVIS_EVENT_TYPE}" == "cron" ) ]]; then
       ./gradlew -PgeckoDriverPlatform=linux64 -PgeckoDriverVersion=0.20.0 build testAll
+    elif [[ "${TRAVIS_JDK_VERSION}" == "oraclejdk9" ]]; then
```
Comment (javabrett) [inline .travis.yml]: Rather than a nested `elif`, could this work near the top of the script, exporting `JDK_JAVA_OPTIONS`?

**Comment by javabrett** [comment]
I merged the testing-container-split change, so good time to rebase this and pick-up that change and resolve conflicts - might give better visibility of test results.

File: .travis.yml (hunk 1 above)
Comment (henrik242) [reply .travis.yml]: Yup, fixed :)

**Comment by boris-petrov** [comment]
Groovy 2.5.0 is out. I see three places in the changes where the 7 dependencies are used - perhaps that could be DRY-ied up?

**Comment by javabrett** [comment]
I've cherry-picked commit 0cf967e9eaf47488d61bdb11c53df5ad56023dbf (`JDK_JAVA_OPTIONS="--add-modules=java.xml.bind"`) early for `master`.

**Comment by boris-petrov** [comment]
@javabrett - I'm not really sure why this is needed? Can you explain or point me to some discussion which I've missed?

**Comment by boris-petrov** [comment]
Sorry, I should have elaborated. I'm talking about the cherry-picked commit `Add bind module when building for jdk9`. I recently played with (and successfully did) migrating our own application to Java 9 and I didn't have to do such a thing and everything was working fine. I still don't understand what needs this and why.

**Comment by henrik242** [comment]
@boris-petrov I don't remember exactly right now, but there was something (probably a transitive dependency?) in the full integration test suite that referenced `javax.annotations.<something>`.  Then we need `--add-modules` to make the suite run.

**Comment by javabrett** [comment]
The Travis build on `master` now has some JDK9 variants running integration tests, so the impact of not having `JDK_JAVA_OPTIONS="--add-modules=java.xml.bind"` is easily tested - I removed it and pushed a test build:

- Commit: https://github.com/javabrett/gretty/commit/ffed798457b5592090e2ddb7af232f8af8b7dc3d
- Travis: https://travis-ci.org/javabrett/gretty/builds/386080081 (three JDK9 builds fail).
- https://travis-ci.org/javabrett/gretty/jobs/386080091#L2094

**Comment by boris-petrov** [comment]
I think I also hit some of these issues. What I did in the end:

1) I had a `jaxb` configuration with dependencies like so:

```gradle
jaxb "javax.xml.bind:jaxb-api:$jaxbVersion"
jaxb "com.sun.xml.bind:jaxb-core:$jaxbVersion"
jaxb "com.sun.xml.bind:jaxb-xjc:$jaxbVersion"
jaxb "com.sun.xml.bind:jaxb-impl:$jaxbVersion"
```

I had to add `jaxb 'javax.activation:activation:1.1.1'` to that.

2) My "main" configuration is called "backend" and I had to add the following two:

```gradle
backendCompileOnly 'javax.annotation:javax.annotation-api:1.3.2'
backendImplementation configurations.jaxb
```

And I think that resolved all such issues. Could something like this be done here?

**Comment by henrik242** [comment]
> backendCompileOnly 'javax.annotation:javax.annotation-api:1.3.2'

Won't this clash with the javax.annotation package ownership for jdk9+, thus requiring `--patch-module` (which is cumbersome...)?

**Comment by boris-petrov** [comment]
I have no idea. Everything works (tests, development, etc) with these changes under JDK9 so I don't think there is a problem. I haven't patched any module.

**Comment by henrik242** [comment]
> I have no idea. Everything works (tests, development, etc) with these changes under JDK9 so I don't think there is a problem.

That's good to know.  Submit a PR if you can :)

**Comment by javabrett** [comment]
Right, the eventual proper fix is to add the external dependencies, possibly for JDK9 builds/executions only.  Raised #62 .  The above is a quick-fix.

**Comment by javabrett** [comment]
@henrik242 As you'll see from the Travis build, we see conflicts between the local Groovy 2.5, and Groovy 2.4 as part of Gradle (now 4.9), when running `groovydoc` tasks:

```
FAILURE: Build completed with 15 failures.
1: Task failed with an exception.
-----------
* What went wrong:
Execution failed for task ':libs:gretty-core:groovydoc'.
> java.lang.NoClassDefFoundError: groovy/text/TemplateEngine
* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.
==============================================================================
```

`groovy/text/TemplateEngine` is I believe new in Groovy 2.5.x, so it bombs-out `groovydoc` running from Gradle.

I think the main motivation for Groovy 2.5 was/is better JDK10 support.  Note that JDK10 is not an LTS so I think we will adopt it in order to absorb incremental changes and for readiness.  Also I just updated master to use `oraclejdk10` instead of `oraclejdk9` and that builds fine.

Gradle master is still on Groovy 2.4.x and I have not researched when they plan to change.

Based on all of the above, unless you have strong feelings otherwise, I think we could park this indefinitely until Gradle moves-up, or we start seeing JDK10 then JDK11 problems, but please let me know your thoughts.

**Comment by javabrett** [comment]
Plenty of interest in getting this merged in discussion over on #96.

Tests have never passed on Travis.  First fix required is that all `groovydoc` tasks fail with `NoClassDefFoundError "groovy/text/TemplateEngine"`.  That needs patching in accordance with:

- http://groovy.329449.n5.nabble.com/Groovydoc-task-failing-with-2-5-1-td5753387.html
- https://discuss.gradle.org/t/gradle-task-groovydoc-failing-with-noclassdeffounderror/2541/6

**Comment by javabrett** [comment]
I think this is pending a change to Gretty to _build_ using Gradle 5. There are one or two known issues for that.

**Comment by boris-petrov** [comment]
@javabrett - I would like to get this going. What do we need to fix? I tried updating the local Gradle to 5.2.1 and running `./gradlew build` and it seems to work fine. Am I missing something? How to make something break so I can try fixing it?

**Comment by javabrett** [comment]
@boris-petrov The change has never passed CI on Travis.  I will try and come back to this over the coming days to rebase the change and make any other updates, and try and explain why it doesn't currently pass.

**Comment by boris-petrov** [comment]
Thanks! I think also Gradle has to be updated to 5.2.1 in this PR so that the Groovy version there is also 2.5.4.

**Comment by javabrett** [comment]
One of the problems should be clear from https://travis-ci.org/gretty-gradle-plugin/gretty/jobs/499155615 - I think I remember there being a `groovydoc` version clash, which leads to a bunch of `java.lang.NoClassDefFoundError: groovy/text/TemplateEngine` during the build.

**Comment by javabrett** [comment]
Looks like this can probably get through with a change to build the Gretty project itself with Gradle 5.2.x.

**Comment by boris-petrov** [comment]
@javabrett - are you going to update it in master or in this PR? Or do you have something else in mind?

**Comment by javabrett** [comment]
@boris-petrov let's do the Gradle 5.x build update in a separate branch/PR.  I pushed branch https://github.com/gretty-gradle-plugin/gretty/tree/gradle-5-2-1 yesterday, no PR yet.

There's some work to be done in Travis config for this ... there's some hackery in-place at the moment to try and allow the build and integration-tests to run on different Gradle versions, but there's probably some issue with that.  Travis build https://travis-ci.org/gretty-gradle-plugin/gretty/builds/499526626 - only the Gradle 5.2.1 test-run completed.

**Comment by boris-petrov** [comment]
@javabrett - well, the build timeouts for some reason... we could always bump the major version of Gretty and support only 5+ from that point on. As a last resort I mean if we cannot figure out why Gradle 4 blocks.

**Comment by boris-petrov** [comment]
@javabrett, @henrik242 - I don't understand what's going on. Locally when running tests I'm getting the `java.lang.LinkageError` that is in the [JDK 8 Travis build](https://travis-ci.org/gretty-gradle-plugin/gretty/jobs/611785318#L1424), while in the [JDK 11 Travis build](https://travis-ci.org/gretty-gradle-plugin/gretty/jobs/611785317#L1456) the JVM segfaults?! Any ideas?

**Comment by boris-petrov** [comment]
@javabrett, @henrik242 - I fixed the build. Can you please review the changes?

@javabrett - if it's OK, we can merge this and release a Gretty 3.0 with the latest changes? Dropping support for Gradle < 5.

**Pull request merged**